chore(public-docsite-v9): updates slot documentation (#28822)
microsoft/fluentui: apps/public-docsite-v9/src/Concepts/Slots/Slots.stories.mdx

Hooks like `useButtonStyles` create classes and conditionally apply them based on the input state. Classes are typically created and applied with `makeStyles` and `mergeClasses`.
For instance, if a component is disabled then disabled styles are added.

### Declaring components slots in the state

The `Button` component defines its slots `root` and `icon` in the state.

```ts
const useButton_unstable = (props: ButtonProps, ref: React.Ref<HTMLButtonElement | HTMLAnchorElement>): ButtonState => {
  return {
    root: slot.always({ ...props, ref }, { elementType: 'button' }),
    icon: slot.optional(props.icon, { elementType: 'span' }),
  };
};
```

> `slot.always` creates a slot that will always render, and as such the user may not provide `null` to opt-out of this slot (NonNullable slot). `slot.optional` creates a slot that can be opted out and is not rendered by default, it only renders if `props.icon` is different from `undefined` (Optional slot).

Reviewer: _nit:_

```suggestion
`slot.always` creates a slot that will always render, and as such the user may not provide `null` to opt-out of this slot (NonNullable slot). `slot.optional` creates a slot that can be opted out of and is not rendered by default, it only renders if `props.icon` is different from `undefined` (Optional slot).
```

> `slot.always` creates a slot that will always render, and as such the user may not provide `null` to opt-out of this slot (NonNullable slot). `slot.optional` creates a slot that can be opted out and is not rendered by default, it only renders if `props.icon` is different from `undefined` (Optional slot).

Reviewer 2: nit: probably a list will be more readable

```suggestion
- `slot.always` creates a slot that will always render, and as such the user may not provide `null` to opt-out of this slot (NonNullable slot).
- `slot.optional` creates a slot that can be opted out of and is not rendered by default, it only renders if `props.icon` is different from `undefined` (Optional slot).
```

> Both `slot.always` and `slot.optional` methods will create a slot definition that can be consumed by the render method to properly render a slot, these methods ensure the local logic provided by the state hook will remain on the slots internals.

Reviewer: _nit:_

```suggestion
Both `slot.always` and `slot.optional` methods will create a slot definition that can be consumed by the render method to properly render a slot, these methods ensure the local logic provided by the state hook will remain on the slots' internals.
```

### Rendering components with slots

- `renderButton` takes in `ButtonState` and renders content.

```tsx
/** @jsxRuntime classic */

> /** @jsx createElement */

Reviewer 2: I personally don't want to expose old pragma, but as the example exists. Let's do it 😿

Author: I'll see to update the doc once again once we got the pragma removed with this https://github.com/microsoft/fluentui/pull/28810

// createElement custom JSX pragma is required to support slot creation
import { createElement } from '@fluentui/react-jsx-runtime';

> import { assertSlots } from '@fluentui/react-utilities';
> const renderButton_unstable = (state: ButtonState) => {

Reviewer 2: ```suggestion
import { assertSlots } from '@fluentui/react-utilities';

const renderButton_unstable = (state: ButtonState) => {
```

nit

  const { iconOnly, iconPosition } = state;

  assertSlots<ButtonSlots>(state);

  return (
    <state.root>
      {iconPosition !== 'after' && state.icon && <state.icon />}
      {!iconOnly && state.root.children}
      {iconPosition === 'after' && state.icon && <state.icon />}
    </state.root>
  );
};
```

> The `assertSlots` method ensures the state has the expected slots. It also provides strong typings to ensure slots are renderable, meaning you can simply `<state.slot/>` and all properties provided to a `slot` creation is already baked into it.

Reviewer: _nit:_ wording

```suggestion
The `assertSlots` method ensures the state has the expected slots. It also provides strong typings to ensure slots are renderable, meaning you can simply use `<state.slot />` and all properties provided to a `slot` creation will be already baked into it.
```
